File: vpn-core-android/vpn-core/src/main/java/me/proton/vpn/core/service/usecases/NetworkObserver.kt
/*
 * Copyright (c) 2025 Proton AG
 *
 * This file is part of ProtonVPN.
 *
 * ProtonVPN is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ProtonVPN is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ProtonVPN.  If not, see <https://www.gnu.org/licenses/>.
 */

package me.proton.vpn.core.service.usecases

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.net.ConnectivityManager
import android.net.LinkProperties
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkInfo
import android.net.NetworkRequest
import android.os.Build
import android.os.Handler
import android.os.HandlerThread
import androidx.annotation.RequiresApi
import androidx.core.content.getSystemService
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import me.proton.vpn.core.api.Logger
import uniffi.protun.ConnectivityEvent
import uniffi.protun.LogLevel

/**
 * Emits [ConnectivityEvent] for VPN's default underlying network, which VPN's protected sockets
 * should bind to.
 *
 * Implementations will use dedicated HandlerThread("NetworkObserver") to run network callbacks on.
 */
internal interface NetworkObserver {

    /**
     * Whether an underlying network is currently available. A snapshot, read once when starting a
     * connection to seed its initial config; ongoing changes are delivered via [events].
     */
    val isNetworkAvailable: Boolean

    /**
     * Connectivity transitions of the active underlying network ([ConnectivityEvent.UP] /
     * [ConnectivityEvent.DOWN] / [ConnectivityEvent.NETWORK_SWITCH]), to forward to the connection.
     */
    val events: SharedFlow<ConnectivityEvent>
}

internal fun NetworkObserver(appContext: Context, logger: Logger): NetworkObserver {
    val connectivityManager = appContext.getSystemService<ConnectivityManager>()
        ?: error("ConnectivityManager is not available.")
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
        BestMatchingNetworkObserver(connectivityManager, logger)
    } else {
        LegacyNetworkObserver(appContext, connectivityManager, logger)
    }
}

/**
 * API 31+ implementation. [ConnectivityManager.registerBestMatchingNetworkCallback] tracks the
 * single best network matching the request (here: an internet, non-VPN network), letting the
 * framework apply its own scoring - which should match what [android.net.VpnService.protect] binds to.
 */
@RequiresApi(31)
private class BestMatchingNetworkObserver(
    private val connectivityManager: ConnectivityManager,
    private val logger: Logger,
) : NetworkObserver {

    // The current best network reported by the framework (may not be validated yet).
    private var bestNetwork: Network? = null
    // The last usable (validated) network we reported; the identity used to detect transitions.
    private var activeNetwork = MutableStateFlow<Network?>(null)

    private val _events = MutableSharedFlow<ConnectivityEvent>(
        extraBufferCapacity = 1,
        onBufferOverflow = BufferOverflow.DROP_OLDEST,
    )
    override val events: SharedFlow<ConnectivityEvent> = _events.asSharedFlow()
    override val isNetworkAvailable get() = activeNetwork.value != null

    private val networkCallback = object : ConnectivityManager.NetworkCallback() {

        override fun onAvailable(network: Network) {
            bestNetwork = network
            evaluate(network, connectivityManager.getNetworkCapabilities(network))
        }

        override fun onCapabilitiesChanged(network: Network, capabilities: NetworkCapabilities) {
            if (network == bestNetwork)
                evaluate(network, capabilities)
        }

        override fun onLost(network: Network) {
            if (network == bestNetwork) {
                bestNetwork = null
                update(null)
            }
        }
    }

    private fun evaluate(network: Network, capabilities: NetworkCapabilities?) {
        if (capabilities != null && capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)) {
            if (network != activeNetwork.value) {
                logger.logNetwork(network, capabilities, connectivityManager.getLinkProperties(network))
                update(network)
            }
        } else {
            update(null)
        }
    }

    private fun update(newActiveNetwork: Network?) {
        val oldActiveNetwork = activeNetwork.value
        val event = when {
            oldActiveNetwork == null && newActiveNetwork != null -> ConnectivityEvent.UP
            oldActiveNetwork != null && newActiveNetwork == null -> ConnectivityEvent.DOWN
            oldActiveNetwork != newActiveNetwork -> ConnectivityEvent.NETWORK_SWITCH
            else -> null
        }
        activeNetwork.value = newActiveNetwork
        if (event != null)
            _events.tryEmit(event)
    }

    init {
        // registerBestMatchingNetworkCallback requires a Handler; use a dedicated background
        // thread so callbacks don't run on the main thread.
        val handlerThread = HandlerThread("NetworkObserver").apply { start() }
        val request = NetworkRequest.Builder()
            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .addCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN)
            .build()
        connectivityManager.registerBestMatchingNetworkCallback(request, networkCallback, Handler(handlerThread.looper))
    }
}

/**
 * API 25-30 implementation based on deprecated [NetworkInfo].
 */
@Suppress("DEPRECATION")
private class LegacyNetworkObserver(
    appContext: Context,
    private val connectivityManager: ConnectivityManager,
    private val logger: Logger,
) : NetworkObserver {

    private val _events = MutableSharedFlow<ConnectivityEvent>(
        extraBufferCapacity = 1,
        onBufferOverflow = BufferOverflow.DROP_OLDEST,
    )
    override val events: SharedFlow<ConnectivityEvent> = _events.asSharedFlow()

    private var lastActiveNetworkInfo: NetworkInfo? = null

    override val isNetworkAvailable get() = lastActiveNetworkInfo?.isConnected == true

    private val receiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (ConnectivityManager.CONNECTIVITY_ACTION != intent.action) return

            val eventInfo = intent.extras?.get(ConnectivityManager.EXTRA_NETWORK_INFO) as? NetworkInfo
            if (eventInfo != null && eventInfo.type != ConnectivityManager.TYPE_VPN) {
                val newActiveNetworkInfo = connectivityManager.activeNetworkInfo

                // Ignore if connectivity action is not about connecting, but an active network is
                // present to avoid duplicated switch events. On some network switches we'll receive
                // DISCONNECTED for the old network first followed by CONNECTED for the new one.
                if (eventInfo.state != NetworkInfo.State.CONNECTED && newActiveNetworkInfo != null)
                    return

                val oldActiveNetworkInfo = lastActiveNetworkInfo
                updateNetworkInfo(newActiveNetworkInfo)
                val event = when {
                    oldActiveNetworkInfo != null && newActiveNetworkInfo == null -> ConnectivityEvent.DOWN
                    oldActiveNetworkInfo == null && newActiveNetworkInfo != null -> ConnectivityEvent.UP
                    oldActiveNetworkInfo != null && newActiveNetworkInfo != null -> ConnectivityEvent.NETWORK_SWITCH
                    else -> null
                }
                event?.let { _events.tryEmit(it) }
            }
        }
    }

    init {
        val handlerThread = HandlerThread("NetworkObserver").apply { start() }
        val handler = Handler(handlerThread.looper)
        handler.post {
            updateNetworkInfo(connectivityManager.activeNetworkInfo)
        }
        appContext.registerReceiver(
            receiver,
            IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION),
            null,
            handler
        )
    }

    private fun updateNetworkInfo(newNetworkInfo: NetworkInfo?) {
        logger.log(LogLevel.INFO, "NetworkObserver: new network=$newNetworkInfo")
        lastActiveNetworkInfo = newNetworkInfo
    }
}

fun Logger.logNetwork(
    network: Network,
    capabilities: NetworkCapabilities,
    linkProperties: LinkProperties?
) {
    val isVpn = capabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN)
    val type = when {
        capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> "WiFi"
        capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> "Mobile"
        capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> "Ethernet"
        else -> "Other"
    }
    log(LogLevel.INFO, "NetworkObserver: network validated $network $type, VPN: $isVpn, addresses: ${linkProperties?.linkAddresses}")
}

File: src/utils/windows/network_events/network_observer.rs
// Copyright (c) 2026 Proton AG
//
// This file is part of ProtonVPN.
//
// ProtonVPN is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// ProtonVPN is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with ProtonVPN.  If not, see <https://www.gnu.org/licenses/>.

use std::net::IpAddr::{V4, V6};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;
use std::sync::mpsc::{self, Sender};

use crate::api::connection::ConnectivityEvent;
use crate::connection::windows::helpers::internet_interface_finder::{InterfaceFinderLogMode, Ipv4InternetInterface, Ipv6InternetInterface, get_internet_interfaces};
use crate::utils::windows::network_events::network_event_listener::NetworkChangesListener;
use crate::utils::windows::network_events::network_observer_events::{NetworkInterface, NetworkObserverEvent};
use crate::utils::windows::debouncer::Debouncer;

const DEBOUNCING_DELAY: Duration = Duration::from_millis(500);

pub(crate) type OnNetworkObserverTriggerType = Box<dyn Fn(ConnectivityEvent) + Send + Sync + 'static>;

pub struct NetworkObserver {
    _network_changes_listener: NetworkChangesListener,
    current_interface_mutex: Arc<Mutex<Option<NetworkInterface>>>,
    network_event_sender: Sender<NetworkObserverEvent>,
}

impl NetworkObserver {
    pub fn start_new_thread(trigger_func: OnNetworkObserverTriggerType) -> NetworkObserver {
        let (network_event_sender, network_event_receiver) = mpsc::channel::<NetworkObserverEvent>();

        let network_changes_listener: NetworkChangesListener = NetworkChangesListener::start(network_event_sender.clone());

        let current_interface_mutex: Arc<Mutex<Option<NetworkInterface>>> = Arc::new(Mutex::new(None));
        let debouncer_current_interface_mutex = Arc::clone(&current_interface_mutex);
        
        thread::spawn(move || {
            Debouncer::start(network_event_receiver, DEBOUNCING_DELAY, move |_| {
                let current_interface: Option<NetworkInterface> = if let Ok(current_interface_lock) = debouncer_current_interface_mutex.lock() {
                    current_interface_lock.clone()
                } else {
                    return;
                };
                handle_debouncer_trigger(current_interface, &trigger_func);
            });
        });
        
        NetworkObserver {
            _network_changes_listener: network_changes_listener,
            current_interface_mutex: current_interface_mutex,
            network_event_sender: network_event_sender,
        }
    }

    pub fn send_current_interface(&self, current_interface: Option<NetworkInterface>) {
        if let Ok(mut current_interface_lock) = self.current_interface_mutex.lock() {
            let current_interface_clone = current_interface.clone();
            *current_interface_lock = current_interface;
            let _ = self.network_event_sender.send(NetworkObserverEvent::InterfaceSelected(current_interface_clone));
        }
    }
}

fn handle_debouncer_trigger(current_interface: Option<NetworkInterface>, trigger_func: &OnNetworkObserverTriggerType) {
    let (new_ipv4_interface, new_ipv6_interface) = get_internet_interfaces(&InterfaceFinderLogMode::ErrorsOnly);

    if new_ipv4_interface.is_none() && new_ipv6_interface.is_none() {
        handle_no_internet_connectivity(current_interface, trigger_func);
    }
    else if new_ipv6_interface.is_none() {
        handle_ipv4_internet_connectivity(new_ipv4_interface.unwrap(), current_interface, trigger_func);
    }
    else if new_ipv4_interface.is_none() {
        handle_ipv6_internet_connectivity(new_ipv6_interface.unwrap(), current_interface, trigger_func);
    }
    else {
        handle_ipv4_and_ipv6_internet_connectivity(new_ipv4_interface.unwrap(), new_ipv6_interface.unwrap(), current_interface, trigger_func);
    }
}

fn handle_no_internet_connectivity(current_interface: Option<NetworkInterface>, trigger_func: &OnNetworkObserverTriggerType) {
    match current_interface {
        Some(current_interface) => {
            log::warn!("Internet connectivity lost (Current interface: {current_interface})");
            trigger_func(ConnectivityEvent::Down);
        },
        None => log::debug!("No internet connectivity"),
    }
}

fn handle_ipv4_internet_connectivity(new_ipv4_interface: Ipv4InternetInterface,
    current_interface: Option<NetworkInterface>, trigger_func: &OnNetworkObserverTriggerType) {
    match current_interface {
        Some(current_interface) => {
            if current_interface.index == new_ipv4_interface.interface_index &&
                let V4(current_interface_ipv4_addr) = current_interface.ip_addr && current_interface_ipv4_addr == new_ipv4_interface.local_ip
            {
                log::debug!("Best IPv4 internet interface already selected ({current_interface})");
            }
            else {
                log::warn!("Best internet interface changed. Requesting switch. \
                    (Current interface: {current_interface}) \
                    (Suggested interface: {new_ipv4_interface})");
                trigger_func(ConnectivityEvent::NetworkSwitch);
            }
        },
        None => handle_internet_connectivity_regained(trigger_func),
    }
}

fn handle_internet_connectivity_regained(trigger_func: &OnNetworkObserverTriggerType) {
    log::warn!("Internet connectivity regained");
    trigger_func(ConnectivityEvent::Up);
}

fn handle_ipv6_internet_connectivity(new_ipv6_interface: Ipv6InternetInterface,
    current_interface: Option<NetworkInterface>, trigger_func: &OnNetworkObserverTriggerType) {
    match current_interface {
        Some(current_interface) => {
            if current_interface.index == new_ipv6_interface.interface_index &&
                let V6(current_interface_ipv6_addr) = current_interface.ip_addr && current_interface_ipv6_addr == new_ipv6_interface.local_ip
            {
                log::debug!("Best IPv6 internet interface already selected ({current_interface})");
            }
            else {
                log::warn!("Best internet interface changed. Requesting switch. \
                    (Current interface: {current_interface}) \
                    (Suggested interface: {new_ipv6_interface})");
                trigger_func(ConnectivityEvent::NetworkSwitch);
            }
        },
        None => handle_internet_connectivity_regained(trigger_func),
    }
}

fn handle_ipv4_and_ipv6_internet_connectivity(new_ipv4_interface: Ipv4InternetInterface, new_ipv6_interface: Ipv6InternetInterface,
    current_interface: Option<NetworkInterface>, trigger_func: &OnNetworkObserverTriggerType) {
    match current_interface {
        Some(current_interface) => {
            if current_interface.index == new_ipv4_interface.interface_index &&
                let V4(current_interface_ipv4_addr) = current_interface.ip_addr && current_interface_ipv4_addr == new_ipv4_interface.local_ip
            {
                log::debug!("Best IPv4 internet interface already selected. IPv6 is available. ({current_interface})");
            }
            else if current_interface.index == new_ipv6_interface.interface_index &&
                let V6(current_interface_ipv6_addr) = current_interface.ip_addr && current_interface_ipv6_addr == new_ipv6_interface.local_ip
            {
                log::debug!("Best IPv6 internet interface already selected. IPv4 is available. ({current_interface})");
            }
            else {
                log::warn!("Best internet interface changed. Requesting switch. \
                    (Current interface: {current_interface}) \
                    (Suggested IPv4 interface: {new_ipv4_interface}) \
                    (Suggested IPv6 interface: {new_ipv6_interface})");
                trigger_func(ConnectivityEvent::NetworkSwitch);
            }
        },
        None => handle_internet_connectivity_regained(trigger_func),
    }
}